File: MaWPhotos/src/main/java/us/mikeandwan/photos/services/PhotoApi.java
package us.mikeandwan.photos.services;

import java.util.List;

import okhttp3.MultipartBody;
import retrofit2.Call;
import retrofit2.http.Body;
import retrofit2.http.GET;
import retrofit2.http.Multipart;
import retrofit2.http.POST;
import retrofit2.http.Part;
import retrofit2.http.Path;
import us.mikeandwan.photos.models.Category;
import us.mikeandwan.photos.models.Comment;
import us.mikeandwan.photos.models.CommentPhoto;
import us.mikeandwan.photos.models.ExifData;
import us.mikeandwan.photos.models.FileOperationResult;
import us.mikeandwan.photos.models.Photo;
import us.mikeandwan.photos.models.PhotoAndCategory;
import us.mikeandwan.photos.models.RatePhoto;
import us.mikeandwan.photos.models.Rating;


interface PhotoApi {
    @GET("photos/getRecentCategories/{sinceId}")
    Call<List<Category>> getRecentCategories(@Path("sinceId") int sinceId);

    @GET("photos/getPhotoExifData/{photoId}")
    Call<ExifData> getExifData(@Path("photoId") int photoId);

    @GET("photos/getRandomPhoto")
    Call<PhotoAndCategory> getRandomPhoto();

    @GET("photos/getCommentsForPhoto/{photoId}")
    Call<List<Comment>> getComments(@Path("photoId") int photoId);

    @GET("photos/getRatingForPhoto/{photoId}")
    Call<Rating> getRatings(@Path("photoId") int photoId);

    @GET("photos/getPhotosByCategory/{categoryId}")
    Call<List<Photo>> getPhotosByCategory(@Path("categoryId") int categoryId);

    @POST("photos/ratePhoto")
    Call<Float> ratePhoto(@Body RatePhoto rating);

    @POST("photos/addCommentForPhoto")
    Call<Boolean> addCommentForPhoto(@Body CommentPhoto commentPhoto);

    @Multipart
    @POST("upload/upload")
    Call<FileOperationResult> uploadFile(@Part MultipartBody.Part file);
}

File: MaWPhotos/src/main/java/us/mikeandwan/photos/services/PhotoApi.kt
package us.mikeandwan.photos.services

import okhttp3.MultipartBody
import retrofit2.Call
import retrofit2.http.*
import us.mikeandwan.photos.models.*

internal interface PhotoApi {
    @GET("photo-categories/recent/{sinceId}")
    fun getRecentCategories(@Path("sinceId") sinceId: Int): Call<ApiCollection<Category>>

    @GET("photos/{photoId}/exif")
    fun getExifData(@Path("photoId") photoId: Int): Call<ExifData>

    @GET("photos/random")
    fun getRandomPhoto(): Call<Photo>

    @GET("photos/random/{count}")
    fun getRandomPhotos(@Path("count") count: Int): Call<ApiCollection<Photo>>

    @GET("photos/{photoId}/comments")
    fun getComments(@Path("photoId") photoId: Int): Call<ApiCollection<Comment>>

    @GET("photos/{photoId}/rating")
    fun getRatings(@Path("photoId") photoId: Int): Call<Rating>

    @GET("photo-categories/{categoryId}/photos")
    fun getPhotosByCategory(@Path("categoryId") categoryId: Int): Call<ApiCollection<Photo>>

    @PATCH("photos/{photoId}/rating")
    fun ratePhoto(@Path("photoId") photoId: Int, @Body rating: RatePhoto): Call<Rating>

    @POST("photos/{photoId}/comments")
    fun addCommentForPhoto(
        @Path("photoId") photoId: Int,
        @Body commentPhoto: CommentPhoto
    ): Call<ApiCollection<Comment>>

    @Multipart
    @POST("upload/upload")
    fun uploadFile(@Part file: MultipartBody.Part): Call<FileOperationResult>
}